walk/walk.go

```diff
@@ -384,22 +387,37 @@ func buildTrie(c *config.Config, updateRels *UpdateFilter, isIgnored isIgnoredFu
 	// An error group to handle error propagation
 	eg := errgroup.Group{}
 	eg.Go(func() error {
-		return walkDir(c.RepoRoot, "", &eg, limitCh, updateRels, isIgnored, trie)
+		return walkDir(c, c.ValidBuildFileNames, "", &eg, limitCh, updateRels, isIgnored, trie)
 	})
 
 	return trie, eg.Wait()
 }
 
 // walkDir recursively and concurrently descends into the 'rel' directory and builds a trie
-func walkDir(root, rel string, eg *errgroup.Group, limitCh chan struct{}, updateRels *UpdateFilter, isIgnored isIgnoredFunc, trie *pathTrie) error {
+func walkDir(c *config.Config, validBuildFileNames []string, rel string, eg *errgroup.Group, limitCh chan struct{}, updateRels *UpdateFilter, isIgnored isIgnoredFunc, trie *pathTrie) error {
 	limitCh <- struct{}{}
 	defer (func() { <-limitCh })()
 
-	entries, err := os.ReadDir(filepath.Join(root, rel))
+	// Absolute path to the directory being visited
+	dir := filepath.Join(c.RepoRoot, rel)
+
+	entries, err := os.ReadDir(dir)
 	if err != nil {
 		return err
 	}
 
+	trie.build, trie.buildFileErr = loadBuildFile(c, validBuildFileNames, rel, dir, entries)
+
+	if trie.build != nil {
+		// Update the `validBuildFileNames` config while traversing the tree.
+		for _, d := range trie.build.Directives {
+			if d.Key == "build_file_name" {
+				validBuildFileNames = strings.Split(d.Value, ",")
+				break
+			}
+		}
+	}
+
 	for _, entry := range entries {
```

Comment:
With a bit more work we can now respect `# gazelle:exclude` here to skip any file/dirs that are excluded. Hopefully it would be an `isExcluded(entryPath)` alongside the `isIgnored(entryPath)` a few lines down within this loop.

Reply:
Here's a poc of that: https://github.com/jbedard/bazel-gazelle/compare/loadbuild-concurrent...jbedard:bazel-gazelle:fs-exclude